// @electron-forge/web-multi-logger/dist/Logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tab = void 0;
const node_path_1 = __importDefault(require("node:path"));
const express_1 = __importDefault(require("express"));
const express_ws_1 = __importDefault(require("express-ws"));
const Tab_1 = __importDefault(require("./Tab"));
exports.Tab = Tab_1.default;
class Logger {
    constructor(port = 9000) {
        this.port = port;
        this.app = (0, express_1.default)();
        this.tabs = [];
        this.server = null;
        this.registerRoutes();
    }
    registerRoutes() {
        this.ws = (0, express_ws_1.default)(this.app);
        this.app.get('/rest/tabs', (_req, res) => res.json(this.tabs));
        this.app.use('/xterm/addons/fit', express_1.default.static(node_path_1.default.dirname(require.resolve('xterm-addon-fit'))));
        this.app.use('/xterm/addons/search', express_1.default.static(node_path_1.default.dirname(require.resolve('xterm-addon-search'))));
        this.app.use('/xterm', express_1.default.static(node_path_1.default.resolve(require.resolve('xterm'), '../..')));
        this.app.use(express_1.default.static(node_path_1.default.resolve(__dirname, '..', 'static')));
        this.ws.app.ws('/sub', () => {
            // I assume this endpoint is just a no-op needed for some reason.
        });
    }
    /**
     * Creates a new tab with the given name, the name should be human readable
     * it will be used as the tab title in the front end.
     */
    createTab(name) {
        const tab = new Tab_1.default(name, this.ws);
        this.tabs.push(tab);
        return tab;
    }
    /**
     * Start the HTTP server hosting the web UI.
     *
     * @returns the port number
     */
    start() {
        return new Promise((resolve) => {
            this.server = this.app.listen(this.port, () => resolve(this.port));
        });
    }
    /**
     * Stop the HTTP server hosting the web UI
     */
    stop() {
        if (this.server)
            this.server.close();
    }
}
exports.default = Logger;
