// @electron-forge/web-multi-logger/dist/Tab.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Log_1 = __importDefault(require("./Log"));
let idCounter = 1;
class Tab {
    constructor(name, ws) {
        this.name = name;
        this.ws = ws;
        this.logs = [];
        this.id = idCounter;
        idCounter += 1;
    }
    /**
     * Log a line to the web UI, a new line is automatically appended to the line
     */
    log(line) {
        const log = new Log_1.default(line, new Date());
        this.logs.push(log);
        for (const client of this.ws.getWss().clients) {
            client.send(JSON.stringify({
                tab: this.id,
                payload: log,
            }));
        }
    }
    toJSON() {
        return {
            id: this.id,
            name: this.name,
            logs: this.logs,
        };
    }
}
exports.default = Tab;
